NAS-108917 / 12.0 / source3/scripts/wsdd.py - gracefully handle oddly formated messages

anodos325 opened the pull request:
User-provided packet captures have show that some Windows 10 clients
may send xml that python element tree fails to parse correctly.

This PR normalizes the string we receive and also catches cases
where required elements (such as header or message id) are missing.
For now we'll log to a file so that I can keep track of this. Issue
manifests itself as the TrueNAS server disappearing from "Network
Neighborhood" on Windows clients when the server receives the
infelicitous message.

themylogin commented:
@anodos325 just wondering, what is an example of such an xml?

anodos325 commented:
> @anodos325 just wondering, what is an example of such an xml?

```'<?xml version="1.0" encoding="UTF-8"?><e:Envelope xmlns:e="http://www.w3.org/2003/05/soap-envelope"xmlns:w="http://schemas.xmlsoap.org/ws/2004/08/addressing"xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery"xmlns:tds="https://www.onvif.org/ver10/device/wsdl/devicemgmt.wsdl"xmlns:dn="http://www.onvif.org/ver10/network/wsdl"><e:Header><w:MessageID>uuid:e50a9480-b702-4b2a-b03a-ecbd5ecbcfef</w:MessageID><w:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</w:To> <w:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</w:Action></e:Header><e:Body><d:Probe><d:Types>dn:NetworkVideoTransmitter</d:Types></d:Probe></e:Body></e:Envelope>'```

in this case ElementTree.from_string() fails due to lack of space between terminating double-quote and "xmlns". This is a rather hackish workaround, but I moved into a separate function in case time shows additional normalization is required before using the python library to convert into an element tree.

themylogin commented:
Is it always so small or it can be big? I see you're logging a complete message, twice (original and processed). Anyway, I would have tried to be more specific, e.g. to fix `"xmlns:` instead or just `"x`

anodos325 committed:
source3/scripts/wsdd.py - gracefully handle oddly formated messages

User-provided packet captures have show that some Windows 10 clients
may send xml that python element tree fails to parse correctly.
This PR normalizes the string we receive and also catches cases
where required elements (such as header or message id) are missing.
For now we'll log to a file so that I can keep track of this. Issue
manifests itself as the TrueNAS server disappearing from "Network
Neighborhood" on Windows clients when the server receives the
infelicitous message.

anodos325 commented:
> Is it always so small or it can be big? I see you're logging a complete message, twice (original and processed). Anyway, I would have tried to be more specific, e.g. to fix `"xmlns:` instead or just `"x`

For posterity in case others look at this convo in the future: the maximum length of message I've seen in pcaps in testing is around 1400 and so rather much to log twice. Protocol documentation in this case is covered in the Microsoft document MS-PBSD and computers in particular in section 2.2 of the aforementioned document.
